### PR TITLE
SQL Parser

### PR DESCRIPTION
**Pull request by AlexValder**

### CODE REVIEW

**Review by ArthurMamedov — APPROVED** [review]

File: DBLab2/DBLab2/ConsoleController/SqlCommands/SqlSelect.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+using System.Collections.Generic;
+
+namespace DBLab2.ConsoleController.SqlCommands {
+    public sealed class SqlSelect : SqlCommand {
+        public List<string>? Fields { get; }
```
Comment (ArthurMamedov) [inline DBLab2/DBLab2/ConsoleController/SqlCommands/SqlSelect.cs:5]: The same irritating question mark.

File: DBLab2/DBLab2/ConsoleController/SqlCommands/SqlSelect.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+        public string Table { get; }
```
Comment (ArthurMamedov) [inline DBLab2/DBLab2/ConsoleController/SqlCommands/SqlSelect.cs:6]: public new string Table { get; } ?

File: DBLab2/DBLab2/ConsoleController/SqlCommands/SqlInsertInto.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+using System.Collections.Generic;
+
+namespace DBLab2.ConsoleController.SqlCommands {
+    public class SqlInsertInto : SqlCommand {
+        public List<string> Values { get; }
+        public List<string>? Fields { get; }
```
[the remark this answers is not in the document]
Comment (AlexValder) [reply DBLab2/DBLab2/ConsoleController/SqlCommands/SqlInsertInto.cs:6]: To show that this property might be null in ready object... And Values should always have a value...

File: DBLab2/DBLab2/ConsoleController/SqlCommands/SqlUpdate.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+
+namespace DBLab2.ConsoleController.SqlCommands { 
+        public class SqlUpdate : SqlCommand {
```
[the remark this answers is not in the document]
Comment (AlexValder) [reply DBLab2/DBLab2/ConsoleController/SqlCommands/SqlUpdate.cs:8]: 4 spaces?

File: DBLab2/DBLab2/ConsoleController/SqlCommands/SqlSelect.cs (hunk 2 above)
Comment (AlexValder) [reply DBLab2/DBLab2/ConsoleController/SqlCommands/SqlSelect.cs:6]: I'll fix...

File: DBLab2/DBLab2/ConsoleController/SqlCommands/SqlInsertInto.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ArthurMamedov) [reply DBLab2/DBLab2/ConsoleController/SqlCommands/SqlInsertInto.cs:10]: Ok, well...

File: DBLab2/DBLab2/ConsoleController/SqlCommands/SqlUpdate.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ArthurMamedov) [reply DBLab2/DBLab2/ConsoleController/SqlCommands/SqlUpdate.cs]: Yep...

File: DBLab2/DBLab2/ConsoleController/SqlCommands/SqlUpdate.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        public SqlUpdate(in string table, in List<(string, string)> set, in (string, Operation, string) cond)
+            : this(table, set) {
+            Condition = cond;
+        }
+
+        public override string ToString() =>
+            $"UPDATE {Table}\n" +
+            $"SET {GetFieldValuePairs()}\n" +
+            (Condition != null ?
+                $"WHERE {Condition.Value.Item1} {Condition.Value.Item2.ToString()} {Condition.Value.Item3}" :
```
Comment (ArthurMamedov) [inline DBLab2/DBLab2/ConsoleController/SqlCommands/SqlUpdate.cs:27]: Replace with: `$"WHERE \"{Condition.Value.Item1}\" {Extension.ToString(Condition.Value.Item2)} \"{Condition.Value.Item3}\""`

**Review by ArthurMamedov — CHANGES_REQUESTED** [review]

File: DBLab2/DBLab2/ConsoleController/SqlCommands/Operation.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            "=" => Operation.Equal,
+            "<" => Operation.Less,
+            ">" => Operation.Greater,
+            "<=" => Operation.EqualOrLess,
+            ">=" => Operation.EqualOrGreater,
+            "<>" => Operation.NonEqual,
+            _ => throw new NotSupportedException()
+        };
+
+        public static string ToString(this Operation @op) => @op switch {
```
[the remark this answers is not in the document]
Comment (AlexValder) [reply DBLab2/DBLab2/ConsoleController/SqlCommands/Operation.cs:31]: Done

File: DBLab2/DBLab2/ConsoleController/SqlCommands/SqlInsertInto.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+            : this(table, values) {
+            Fields = fields;
+        }
+
+        public override string ToString() =>
+            $"INSERT INTO {Table} {GetFields()}\n" +
+            $"VALUES {GetValues()}";
+
+        private string GetFields() =>
+            Fields != null ? $"({string.Join(',', Fields)})" : "";
```
[the remark this answers is not in the document]
Comment (AlexValder) [reply DBLab2/DBLab2/ConsoleController/SqlCommands/SqlInsertInto.cs:26]: Done

File: DBLab2/DBLab2/ConsoleController/SqlCommands/SqlDelete.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+                throw new ArgumentException("Empty SQL statements", nameof(condition));
+            }
+            
+            Condition = condition;
+        }
+
+        public override string ToString() =>
+            $"DELETE FROM {Table}" +
+            (Condition != null
+                ? $" WHERE {Condition.Value.Item1} {Condition.Value.Item2} {Condition.Value.Item3}"
```
[the remark this answers is not in the document]
Comment (AlexValder) [reply DBLab2/DBLab2/ConsoleController/SqlCommands/SqlDelete.cs:29]: Done

File: DBLab2/DBLab2/ConsoleController/SqlCommands/SqlInsertInto.cs [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+        private string GetValues() => $"({string.Join(',', Values)})";
```
[the remark this answers is not in the document]
Comment (AlexValder) [reply DBLab2/DBLab2/ConsoleController/SqlCommands/SqlInsertInto.cs:28]: Done

File: DBLab2/DBLab2/ConsoleController/SqlCommands/SqlUpdate.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+                ""
+            );
+
+        private string GetFieldValuePairs() => string.Join(", ",
+                Set.Select(pair => $"{pair.Item1} = {pair.Item2}"));
```
[the remark this answers is not in the document]
Comment (AlexValder) [reply DBLab2/DBLab2/ConsoleController/SqlCommands/SqlUpdate.cs:32]: Done

File: DBLab2/DBLab2/ConsoleController/SqlCommands/SqlUpdate.cs (hunk 5 above)
Comment (AlexValder) [reply DBLab2/DBLab2/ConsoleController/SqlCommands/SqlUpdate.cs:27]: Done

**Review by ArthurMamedov — APPROVED** [review]

**Pull request merged**